packages/vit-s/src/hooks.js

```diff
@@ -322,3 +323,106 @@ export function useGetObsInfo(obsType, obsLabelsTypes, obsLabelsData, obsSetsMem
+ *      - Mapped Gene List or object
+ *
+ * @example
+ * // Example with an array of gene strings (featureList):
+ * const geneObject = ["ENSG00000123456.1", "ENSG00000987654.2"];
+ * // Example with an object of genes, used in the tooltip:
+ * const geneObject = {
+ *   "Gene ID": "ENSG00000123456.1",
+ *   "Cell ID": "Cell1234",
+ *   "Marker Gene": "ENSG00000987654.2"
+ * };
+ */
+
+export const useMappedGeneList = (geneObject, featureMap) => {
+  const setWarning = useSetWarning();
+  const [mappedGeneList, setMappedGeneList] = useState(geneObject);
+  const [fetchedGenesList, setFetchedGenesList] = useState(null);
+
+  useEffect(() => {
+    // Assuming featureMap is set for symbolic genes
+    if (featureMap !== undefined) {
+      setMappedGeneList(geneObject);
+      return;
+    }
+    const isEnsembleGene = value => value?.toUpperCase().startsWith('ENSG');
+    const isGeneKey = key => key.toLowerCase().includes('gene');
+    const ensbGenes = Array.isArray(geneObject)
+      ? geneObject?.some(isEnsembleGene)
+      : geneObject && Object.entries(geneObject)?.some(([key, value]) => isGeneKey(key)
+              && isEnsembleGene(value));
+    if (!ensbGenes) return;
+
+    const fetchGeneData = async () => {
+      const controller = new AbortController();
+      const { signal } = controller;
+
+      try {
+        const response = await fetch('https://vitessce-resources.s3.us-east-2.amazonaws.com/genes_filtered.json', { signal });
+        if (!response.ok) {
+          throw new Error(`HTTP error! Status: ${response.status}`);
+        }
+        const data = await response.json();
+        setFetchedGenesList(data);
+      } catch (err) {
+        if (err.name !== 'AbortError') {
+          setWarning(err.message);
+        }
+      }
+
+      return () => controller.abort();
+    };
+
+    fetchGeneData();
+  }, [geneObject, featureMap]);
+
+  useEffect(() => {
+    if (!fetchedGenesList || featureMap !== undefined) return;
+
+    const mapGenes = () => {
+      const isEnsembleGene = value => value?.toUpperCase().startsWith('ENSG');
+      const isGeneKey = key => key.toLowerCase().includes('gene');
+
+      let updatedGeneObject = { ...geneObject };
+
+      if (Array.isArray(geneObject)) {
+        updatedGeneObject = geneObject
+          .map((gene) => {
+            if (isEnsembleGene(gene)) {
+              const trimmedValue = gene.split('.')[0]; // Trim after the dot
+              return fetchedGenesList[trimmedValue];
+            }
+            return null;
+          })
+          .filter(Boolean);
+      } else {
+        Object.entries(geneObject).forEach(([key, value]) => {
+          if (isGeneKey(key) && isEnsembleGene(value)) {
+            const trimmedValue = value.split('.')[0];
+            updatedGeneObject[key] = fetchedGenesList[trimmedValue];
+          }
+        });
+      }
+
+      setMappedGeneList(updatedGeneObject);
+    };
+
+    mapGenes();
+  }, [fetchedGenesList, geneObject, featureMap]);
```

Comment:
I think we could make this a `useMemo` that returns the `updatedGeneObject` instead - this would help avoid the extra `setState` at the end of the effect, which will trigger it to rerender again.

Reply:
Good catch, thanks